File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/Flock.java
package com.bbd.abstractFactoryPattern;

import java.util.ArrayList;
import java.util.List;

public class Flock implements IQuackable
{
	private String _name = "Flock";
	private List<IQuackable> flock;
	private IQuackObservable _observable;

	Flock()
	{
		flock = new ArrayList<IQuackable>();
		this._observable = new Observable(this);
	}

	Flock(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public void add(IQuackable duck)
	{
		flock.add(duck);
	}

	@Override
	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	@Override
	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	@Override
	public String quack()
	{
		StringBuilder sb = new StringBuilder();
		notifyObservers();
		for (IQuackable duck : flock)
		{
			sb.append(duck.quack() + " inside the flock\n");
		}

		return (sb.toString());
	}

	@Override
	public String getName() { return (this._name);}
}

File: Scala/compoundlab/src/main/scala/Flock.scala
import scala.collection.mutable.ArrayBuffer

class Flock extends IQuackable {

	private var name = "Flock"
	private val flock : ArrayBuffer[IQuackable] = ArrayBuffer[IQuackable]()

	override def registerObserver(observer: IObserver): Unit = {
		flock.foreach(duck => {
			duck.registerObserver(observer)
		})
	}

	override def notifyObservers: String = {
		val sb : StringBuilder = new StringBuilder
		flock.foreach(duck => {
			sb.append(duck.notifyObservers + "\n")
		})
		sb.toString().trim
	}

	def add(duck : IQuackable) : Unit = {
		flock.append(duck)
	}

	def quack : String = {
		val sb = new StringBuilder

		flock.foreach(flock => {
			sb.append(flock.quack + " from a flock\n")
		})

		sb.toString()
	}

	override def getName: String = {
		this.name
	}
}
